components/suggest/src/db.rs

```diff
@@ -319,13 +326,21 @@ impl<'a> SuggestDao<'a> {
         let suggestions_limit = &query.limit.unwrap_or(-1);
         let suggestions = self.conn.query_rows_and_then_cached(
             "SELECT s.id, k.rank, s.title, s.url, s.provider, s.score, k.confidence, k.keyword_suffix
-                     FROM suggestions s
-                     JOIN prefix_keywords k ON k.suggestion_id = s.id
-                     WHERE k.keyword_prefix = :keyword_prefix AND s.provider = :provider
-                     ORDER BY s.score DESC
-                     LIMIT :suggestions_limit",
+                FROM suggestions s
+                JOIN prefix_keywords k ON k.suggestion_id = s.id
+                JOIN (
+                SELECT
+                    suggestion_id,
+                    MAX(rank) as highest_rank
+                    FROM prefix_keywords WHERE keyword_prefix = :keyword_prefix AND keyword_suffix LIKE :keyword_suffix || '%' 
+                    GROUP BY suggestion_id
+                ) h ON k.suggestion_id = h.suggestion_id AND k.rank = h.highest_rank
+                WHERE s.provider = :provider
+                ORDER BY s.score DESC
+                LIMIT :suggestions_limit",
```

Comment:
Same here. Since this query is the same as the AMO one -- and probably any suggestion type that uses the `prefix_keywords` table will use it -- is it feasible to factor out either the SQL or the surrounding Rust code into a common helper? Maybe a method that queries the table given a query and provider? That's fine for a follow-up though since the SQL is already duplicated.

Reply:
considering we made changes to the sql statement for pocket and mdn and they slightly differ, i think i'm going to opt to do this as a follow up 🙂